FontBakery dashboard

felipesanches opened the pull request:
This is a work in progress. I'll merge it once it is done.

From time to time I update the live instance with this so that the 'under-construction' dashboard can be seen at http://fontbakery.appspot.com/dashboard

felipesanches commented:
@davelab6 

The dockerized dashboard took longer than I estimated, but finally started working. For now, I hardcoded 4 font projects in the docker-compose setup. I'll figure out a better way of dispatching hundreds of customized worker containers later, for now these 4 are enough for me to test the app locally.

Each worker container checks out a git repo (specified in the REPO_URL environment var passed to the container) and then runs fontbakery-check-ttf.py on all commits in the repo history, storing the full set of check results in a table and cached summaries of the overal score in a separate "cached_stats" table.

The flask webapp queries the database and displays the families with their overal score.

I'm having trouble with the CSS file not being available for the app. Tomorrow I'll see what needs to be done to serve these static files. I briefly saw people suggesting a dedicated nginx container for hosting a flask app static folder. Not sure if that's a good choice, but I'll deal with that tomorrow.

Then I'll create a template for the detailed page. This should be quicker to implement, I hope, given that the database and docker-compose setup is all sorted out now.

cheers,
Felipe

felipesanches commented:
This looks uglier than previous update because it is failing to load the CSS file. I'll fix that tomorrow:

![selecao_715](https://cloud.githubusercontent.com/assets/213676/23245846/4e9c8436-f96e-11e6-96ab-1ab5aa35a983.png)

felipesanches commented:
@davelab6 Details page working!

Next steps: 
* add the actual list of check results
* add the burndown chart maybe also in this same page.
* setup the workers to copy prod files and run fontbakery on those as well
* show the prod versus git upstream data on a separate page on the webapp

![selecao_717](https://cloud.githubusercontent.com/assets/213676/23291145/e3b1a9c2-fa34-11e6-8060-a45b5cbe436f.png)

felipesanches commented:
The per-TTF progress bars seem not very useful (at least not in this case). I may leave them there, or I may use the screen real-estate for things like the burndown chart and the actual check-results.

davelab6 commented:
Throw everything in. We will redesign it later, for now focus on making it work. The screenshot without CSS is totally fine for this stage IMHO.

m4rc1e commented:
Good progress!

I'll get a docker machine up and running and have a play with it soon :-)

I was just expecting json for the time being :-)

felipesanches commented:
![selecao_733](https://cloud.githubusercontent.com/assets/213676/23490021/82890b6e-fed4-11e6-982f-9299d5bdc636.png)

felipesanches commented:
The image from the previous comment is a screenshot of the fontbakery dashboard as of today. I have added the panel with per-fontfile tabs with the individual check-results (ported from the drag-and-drop webapp into a jinja template populated with the data from rethinkdb queries). I have also moved the single-ttf score percentage+progressbar into each of the specific tabs.

I would like to deploy this to Google Container Engine. I have setup my account and created a cluster of 3 nodes (default size, can be changed). I leraned though that the docker-compose.yaml setup file that I am using for testing on my machine is not the same thing that kubernetes expects. In order to deploy on Gogle Container Engine, using the kubectl command line tool, I need to setup a similar, but different yaml file describing the desired deployment. It puzzles me a bit and I'm now trying to figure out exactly how that's supposed to be done.

I found a tool called kompose (available at http://kompose.io/) that tries to make it easier to move from docker-compose into kubernetes resource description files. If failed to do so for fb-dashboard though. As with all automated tools, I already expected it to fail somehow. So now I'm trying to figure out how to set this up manually.

davelab6 commented:
I'm loving it

On Mar 1, 2017 6:26 PM, "Felipe Corrêa da Silva Sanches" <
notifications@github.com> wrote:

> The image from the previous comment is a screenshot of the fontbakery
> dashboard as of today. I have added the panel with per-fontfile tabs with
> the individual check-results (ported from the drag-and-drop webapp into a
> jinja template populated with the data from rethinkdb queries). I have also
> moved the single-ttf score percentage+progressbar into each of the specific
> tabs.
>
> I would like to deploy this to Google Container Engine. I have setup my
> account and created a cluster of 3 nodes (default size, can be changed). I
> leraned though that the docker-compose.yaml setup file that I am using for
> testing on my machine is not the same thing that kubernetes expects. In
> order to deploy on Gogle Container Engine, using the kubectl command line
> tool, I need to setup a similar, but different yaml file describing the
> desired deployment. It puzzles me a bit and I'm now trying to figure out
> exactly how that's supposed to be done.
>
> I found a tool called kompose (available at http://kompose.io/) that
> tries to make it easier to move from docker-compose into kubernetes
> resource description files. If failed to do so for fb-dashboard though. As
> with all automated tools, I already expected it to fail somehow. So now I'm
> trying to figure out how to set this up manually.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlefonts/fontbakery/pull/1209#issuecomment-283536932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAP9y3oFjNtHBQ7HIKEqNkvI017NdEoDks5rhijEgaJpZM4LyOy9>
> .
>

felipesanches commented:
For parallel processing of the full Google Fonts set of font families, I'm considering using a message queue service as described at https://kubernetes.io/docs/user-guide/jobs/work-queue-1/

felipesanches commented:
And for schedulling re-runs of the checks on the full set of families I may use the CronJob feature of Kubernetes, as described at https://kubernetes.io/docs/user-guide/cron-jobs/

felipesanches commented:
This morning I sketched an interesting new view of the date, which will be particularly useful for validading the sanity of the fontbakery test suite itself. It tells uf how often each specific check fails/passes throughout the full set of tested fonts.

![selecao_734](https://cloud.githubusercontent.com/assets/213676/23521245/ee580380-ff5c-11e6-9c90-08ea3a835a5c.png)

felipesanches commented:
The image above is a showcase of it running only on 4 families. But with a full set of families being checked after deploying it to GKE, then we'll be able to notice some patterns such as:

* (A) If a check always fails then maybe the check implementation itself is broken
* (B) If a check passes for all font files of all families, then we either also have a bad check implementation (one that never ever fails) or it's simply an issue that vanished from the Google Fonts project (i.e. all font projects successfully got rid of the problem). The check then remains useful only for a "regression alert" purpose (or for new families yet to be introduced).

felipesanches commented:
Note: there are 15 TTF files being checked, but there's a total of 14 for each check in this table because 1 of the 15 TTFs has a non-canonical name: Cabin-RegularCondensed.ttf

So it errors in that first checks and bails out not running the rest of the testsuite at all.

@davelab6 If you like it, I can create an additional panel for listing all identified non-canonical filenames in the full set of families.

![selecao_736](https://cloud.githubusercontent.com/assets/213676/23521676/b82d2aa4-ff5e-11e6-845e-82c30a92e141.png)

davelab6 commented:
> This morning I sketched an interesting new (and complementary) view of the data in an additional dashboard page, which will be particularly useful for validating the sanity of the fontbakery test suite itself. It tells us how often each specific check fails/passes throughout the full set of tested fonts.

EXCELLENT! The fsType macStyle and `name` table checks will be really helpful for @rsheeter :)

davelab6 commented:
> @davelab6 If you like it, I can create an additional panel for listing all identified non-canonical filenames in the full set of families.

Sure, that's like a "p0" or "ON FIRE! must immediately fix" kind of problem :)

davelab6 commented:
> Sure, that's like a "p0" or "ON FIRE! must immediately fix" kind of problem :)

Each check should probably have a new 'priority level' piece of metadata, similar to 'result type', with 5 values from 0 to 4

felipesanches commented:
@davelab6 I created a very simple system for tagging checks with priority levels. The levels are called TRIVIAL, LOW, NORMAL, IMPORTANT and CRITICAL.

By default checks are implicitely initiated with NORMAL priority level. We can explicitely over-ride that case-by-case by invoking something like fb.set_priority(CRITICAL) right in the beginning of the check implementation.

Now it's up to us to decide which checks are critical. Also, there's no visible indication of priority levels yet. We may use this metadata field for highlighting the critical ones on the web ui. Or maybe for sorting. That's still open for brainstorming :-)

felipesanches commented:
@davelab6 & @m4rc1e 

Phew!!!
I spent the whole of Friday and Saturday frantically dabbling docker/kubernetes lingo on the text terminal !

It's actually pretty cool :-) And powerful!

I've written yaml descriptors for all jobs and services needed for the dashboard. And I replaced the hardcoded set of per-family worker containers, but a single worker container that fetches its attributes (git_repo_url and fontfiles_prefix) from a workqueue. The queue manager is the new piece of the puzzle: I'm using a RabbitMQ message queue server on its own container. And I have added a "job dispatcher" container which is a very simple python script with a hardcoded list of parameters for a selection of 34 font families; The dispatcher job, when started, feeds the queue server which these jobs so that the worker container instances can fetch these for running fontbakery and saving the check results to the database.

By the way, I have setup the worker jobs to run 8 pods in parallel in the cluster, so that processing a large batch of font families can be faster. This is possible in part due to the queue manager taking care of dispatching a unique family descriptor for each worker.

Everything looked quite fine at first. The whole setup, including the messaging for the job queue, works nice locally with docker-compose. So I built the container images of all of these components and uploaded them to the Google Container Engine Registry for docker images.

Then, using the kubectl command-line tool and the yaml descriptions that I had written, I bootstrapped the services, jobs and replication controllers on my GKE cluster. I have written down the sequence of commands in a shell script so that I do not have to memorize it all. The script is called deploy_to_gke.sh and it sits in the root of the fontbakery repo (in this not-yet-merged pull request: https://github.com/googlefonts/fontbakery/pull/1209)

Most things worked ok GKE. But a few other things crashed! :-P So I started inspecting the logs of each pod and noticed that the worker pods were getting a connection timeout when attempting to connect to the rethinkdb database. They also spit a warning that I might be using a too old version of rethinkdb and that only versions newer than 2.3.0 were supported. I immediately noticed that the rethinkdb container image I was using was 1.16.0_1 (https://console.cloud.google.com/gcr/images/google-containers/global/rethinkdb?project=google-containers) which is the newest image available on the Google Containers image registry.

So, I grabbed a Dockerfile rule for a newer rethinkdb image (2.3.5) and built it myself. Then I pushed it into my personal registry on GKE and changed the yaml service descriptor to use that image instead. After changing that, deleting the old-version containers and instantiating a new rethinkdb service/replication-controller on the cluster, I lost the ability to access the web-admin panel (it was working for the older version of rethinkdb, though).

I considered the possibility that perhaps there might be old persistent database data causing some sort of crash due to version incompatibilities between (1.16.0_1 and 2.3.5), but when inspecting the logs it looked like the server did boot up nicely and was readily waiting for connections from peers.

Anyway... out of an abundance of caution I decided to completely delete the cluster and create a new one, starting over from scratch just to be sure. Nice opportunity to validate my notes on that shell script!

The admin panel should be available on the public IP address/port http://104.154.207.226:8080/ (as it was with the older version of rethinkdb) but it silently refuses to load... See the kubectl get services output on the attached screenshot.

I'll rest now and I'll get back to it on Monday.

Happy Hacking,
Felipe Sanches
![selecao_740](https://cloud.githubusercontent.com/assets/213676/23582744/c350b3a8-010f-11e7-97e4-82a91182f132.png)

felipesanches commented:
The dashboard is finally running on Google Container Engine!

Right now it is available at http://104.197.141.134/ and there's also a currently hidden panel at http://104.197.141.134/testsuite (but I may add a link to it on the main dashboard page).

I'll still have to find out how to setup a domain name for services running on specific containers. @davelab6 offered the fontbakery.com domain, which he owns, for us to use here, but I have not set up that yet.

Below are a few screenshots of the dashboard after 21 font-family repos were checked with fontbakery.
![selecao_743](https://cloud.githubusercontent.com/assets/213676/23733252/c6e88474-0456-11e7-9151-7b73ca47221b.png)
![selecao_742](https://cloud.githubusercontent.com/assets/213676/23733253/c6eae6c4-0456-11e7-9015-2fec3ae97684.png)

felipesanches commented:
I have also reorganized the code-tree to host the files and Dockerfile rules for each container in a separate directory. Unfortunately, due https://github.com/docker/docker/issues/2745 the Dockerfile for the worker container could not be place in this organized tree, but needed to go in the root dir of the fontbakery repo so that it could add the check-ttf script into the container.

![selecao_747](https://cloud.githubusercontent.com/assets/213676/23733553/5a3c7fc2-0458-11e7-935d-0a799ec49a55.png)

felipesanches commented:
The job queue mechanism had a few bugs (I was forgetting to signal proper consumption of tasks). Additionally, sometimes I still see worker containers stuck either waiting for a new task or hanging for some yet unknown reason during fontbakery testsuite execution. This needs to be further inspected, because I was not yet able to understand why these get stuck.

I think this pull request can be merged and then I'll continue working on each of these individual issues that I mentioned above.

davelab6 commented:
Yep merge and file issues so I can prioritize please

On Mar 8, 2017 9:45 PM, "Felipe Corrêa da Silva Sanches" <
notifications@github.com> wrote:

> The job queue mechanism had a few bugs (I was forgetting to signal proper
> consumption of tasks). Additionally, sometimes I still see worker
> containers stuck either waiting for a new task or hanging for some yet
> unknown reason during fontbakery testsuite execution. This needs to be
> further inspected, because I was not yet able to understand why these get
> stuck.
>
> I think this pull request can be merged and then I'll continue working on
> each of these individual issues that I mentioned above.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlefonts/fontbakery/pull/1209#issuecomment-285238534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAP9yzkpAqOlSAGRz-IB5TXeRfuOBzh_ks5rj2e6gaJpZM4LyOy9>
> .
>
